Tribler/Core/Upgrade/upgrade.py

```diff
@@ -174,6 +178,34 @@ def upgrade_config_to_71(self):
         self.session.config = convert_config_to_tribler71(self.session.config)
         self.session.config.write()
 
+    def backup_state_directory(self):
+        """
+        Backs up the current state directory if the version in the state directory and in the code is different.
+        """
+        if self.session.config.get_version_backup_enabled() and self.session.config.get_version() \
+                and not self.session.config.get_version() == version_id:
+
+            src_state_dir = self.session.config.get_state_dir()
+            dest_state_dir = self.session.config.get_state_dir(version=self.session.config.get_version())
+
+            # If only there is no tribler config already in the backup directory, then make the current version backup.
+            dest_conf_path = os.path.join(dest_state_dir, 'triblerd.conf')
+            if not os.path.exists(dest_conf_path):
+                # Backup selected directories
+                backup_dirs = [STATEDIR_DB_DIR, STATEDIR_CHECKPOINT_DIR, STATEDIR_WALLET_DIR, STATEDIR_CHANNELS_DIR]
+                src_sub_dirs = os.listdir(src_state_dir)
+                for backup_dir in backup_dirs:
+                    if backup_dir in src_sub_dirs:
+                        dir_copy(os.path.join(src_state_dir, backup_dir), os.path.join(dest_state_dir, backup_dir))
+                    else:
+                        os.makedirs(os.path.join(dest_state_dir, backup_dir))
+
+                # Backup keys and config files
+                backup_files = ['ec_multichain.pem', 'ecpub_multichain.pem', 'ec_trustchain_testnet.pem',
+                                'ecpub_trustchain_testnet.pem', 'triblerd.conf']
```

Comment:
Does this work if one of the files does not exist?

Reply:
Yes.  If the file does not exist it will be created during session initialization and later copied during upgrade.

Reply:
Alright, I thought testnet keys were not generated by default.